File: src/commands/market-data.ts
import type { QuoteImport } from "@/lib/types/quote-import";
import {
  QuoteSummary,
  Asset,
  Quote,
  UpdateAssetProfile,
  MarketDataProviderInfo,
} from "@/lib/types";
import { invokeTauri, logger } from "@/adapters";

// Interface matching the backend struct
export interface MarketDataProviderSetting {
  id: string;
  name: string;
  description: string;
  url: string | null;
  priority: number;
  enabled: boolean;
  logoFilename: string | null;
  lastSyncedAt: string | null;
  lastSyncStatus: string | null;
  lastSyncError: string | null;
}

export const searchTicker = async (query: string): Promise<QuoteSummary[]> => {
  try {
    return invokeTauri("search_symbol", { query });
  } catch (error) {
    logger.error("Error searching for ticker.");
    throw error;
  }
};

export const syncHistoryQuotes = async (): Promise<void> => {
  try {
    await invokeTauri("synch_quotes");
    return;
  } catch (error) {
    logger.error("Error syncing history quotes.");
    throw error;
  }
};

export const getAssetProfile = async (assetId: string): Promise<Asset> => {
  try {
    return invokeTauri("get_asset_profile", { assetId });
  } catch (error) {
    logger.error("Error loading asset data.");
    throw error;
  }
};

export const getAssets = async (): Promise<Asset[]> => {
  try {
    return invokeTauri("get_assets");
  } catch (error) {
    logger.error("Error loading assets.");
    throw error;
  }
};

export const getLatestQuotes = async (symbols: string[]): Promise<Record<string, Quote>> => {
  try {
    return invokeTauri("get_latest_quotes", { symbols });
  } catch (error) {
    logger.error("Error loading latest quotes.");
    throw error;
  }
};

export const updateAssetProfile = async (payload: UpdateAssetProfile): Promise<Asset> => {
  try {
    return invokeTauri("update_asset_profile", { id: payload.symbol, payload });
  } catch (error) {
    logger.error("Error updating asset profile.");
    throw error;
  }
};

export const deleteAsset = async (id: string): Promise<void> => {
  try {
    await invokeTauri("delete_asset", { id });
    return;
  } catch (error) {
    logger.error("Error deleting asset.");
    throw error;
  }
};

export const updateAssetDataSource = async (symbol: string, dataSource: string): Promise<Asset> => {
  try {
    return invokeTauri("update_asset_data_source", { id: symbol, dataSource });
  } catch (error) {
    logger.error("Error updating asset data source.");
    throw error;
  }
};

export const updateQuote = async (symbol: string, quote: Quote): Promise<void> => {
  try {
    return invokeTauri("update_quote", { symbol, quote: quote });
  } catch (error) {
    logger.error("Error updating quote");
    throw error;
  }
};

export const syncMarketData = async (symbols: string[], refetchAll: boolean): Promise<void> => {
  try {
    await invokeTauri("sync_market_data", { symbols, refetchAll });
    return;
  } catch (error) {
    logger.error(`Error refreshing quotes for symbols: ${String(error)}`);
    throw error;
  }
};

export const deleteQuote = async (id: string): Promise<void> => {
  try {
    return invokeTauri("delete_quote", { id });
  } catch (error) {
    logger.error("Error deleting quote");
    throw error;
  }
};

export const getQuoteHistory = async (symbol: string): Promise<Quote[]> => {
  try {
    return await invokeTauri("get_quote_history", { symbol });
  } catch (error) {
    logger.error(`Error fetching quote history for symbol ${symbol}.`);
    throw error;
  }
};

export const getMarketDataProviders = async (): Promise<MarketDataProviderInfo[]> => {
  try {
    return invokeTauri("get_market_data_providers");
  } catch (error) {
    logger.error("Error fetching market data providers.");
    throw error;
  }
};

export const getMarketDataProviderSettings = async (): Promise<MarketDataProviderSetting[]> => {
  try {
    return invokeTauri("get_market_data_providers_settings");
  } catch (error) {
    logger.error("Error fetching market data provider settings.");
    throw error;
  }
};

export const updateMarketDataProviderSettings = async (payload: {
  providerId: string;
  priority: number;
  enabled: boolean;
}): Promise<MarketDataProviderSetting> => {
  try {
    return invokeTauri("update_market_data_provider_settings", payload);
  } catch (error) {
    logger.error("Error updating market data provider settings.");
    throw error;
  }
};

export const importManualQuotes = async (
  quotes: QuoteImport[],
  overwriteExisting: boolean = true,
): Promise<QuoteImport[]> => {
  try {
    return invokeTauri("import_quotes_csv", { quotes, overwriteExisting });
  } catch (error) {
    logger.error("Error importing manual quotes.");
    throw error;
  }
};

File: src-server/src/api/market_data.rs
use std::sync::Arc;

use crate::{
    api::shared::{enqueue_portfolio_job, PortfolioJobConfig},
    error::ApiResult,
    main_lib::AppState,
};
use axum::{
    extract::{Path, Query, State},
    http::StatusCode,
    routing::{delete, get, post, put},
    Json, Router,
};
use wealthfolio_core::quotes::{MarketSyncMode, ProviderInfo, Quote, QuoteImport, QuoteSummary};

async fn get_market_data_providers(
    State(state): State<Arc<AppState>>,
) -> ApiResult<Json<Vec<ProviderInfo>>> {
    let infos = state.quote_service.get_providers_info().await?;
    Ok(Json(infos))
}

async fn get_market_data_provider_settings(
    State(state): State<Arc<AppState>>,
) -> ApiResult<Json<Vec<ProviderInfo>>> {
    let infos = state.quote_service.get_providers_info().await?;
    Ok(Json(infos))
}

#[derive(serde::Deserialize)]
struct ProviderUpdateBody {
    #[serde(rename = "providerId")]
    provider_id: String,
    priority: i32,
    enabled: bool,
}

async fn update_market_data_provider_settings(
    State(state): State<Arc<AppState>>,
    Json(body): Json<ProviderUpdateBody>,
) -> ApiResult<StatusCode> {
    state
        .quote_service
        .update_provider_settings(&body.provider_id, body.priority, body.enabled)
        .await?;
    Ok(StatusCode::NO_CONTENT)
}

#[derive(serde::Deserialize)]
struct SearchQuery {
    query: String,
}

async fn search_symbol(
    State(state): State<Arc<AppState>>,
    Query(q): Query<SearchQuery>,
) -> ApiResult<Json<Vec<QuoteSummary>>> {
    let res = state.quote_service.search_symbol(&q.query).await?;
    Ok(Json(res))
}

#[derive(serde::Deserialize)]
struct QuoteHistoryQuery {
    symbol: String,
}

async fn get_quote_history(
    State(state): State<Arc<AppState>>,
    Query(q): Query<QuoteHistoryQuery>,
) -> ApiResult<Json<Vec<Quote>>> {
    let res = state.quote_service.get_historical_quotes(&q.symbol)?;
    Ok(Json(res))
}

async fn update_quote(
    Path(symbol): Path<String>,
    State(state): State<Arc<AppState>>,
    Json(mut quote): Json<Quote>,
) -> ApiResult<StatusCode> {
    // Ensure asset_id matches path parameter
    quote.asset_id = symbol;
    state.quote_service.update_quote(quote).await?;
    // Manual quote update - no market sync needed
    enqueue_portfolio_job(
        state.clone(),
        PortfolioJobConfig {
            account_ids: None,
            market_sync_mode: MarketSyncMode::None,
            force_full_recalculation: false,
        },
    );
    Ok(StatusCode::NO_CONTENT)
}

async fn delete_quote(
    Path(id): Path<String>,
    State(state): State<Arc<AppState>>,
) -> ApiResult<StatusCode> {
    state.quote_service.delete_quote(&id).await?;
    // Manual quote deletion - no market sync needed
    enqueue_portfolio_job(
        state,
        PortfolioJobConfig {
            account_ids: None,
            market_sync_mode: MarketSyncMode::None,
            force_full_recalculation: false,
        },
    );
    Ok(StatusCode::NO_CONTENT)
}

async fn sync_history_quotes(State(state): State<Arc<AppState>>) -> ApiResult<StatusCode> {
    let result = state.quote_service.resync(None).await?;
    if result.failed > 0 {
        tracing::warn!("resync reported {} failures", result.failed);
    }
    Ok(StatusCode::NO_CONTENT)
}

#[derive(serde::Deserialize)]
struct ImportQuotesBody {
    quotes: Vec<QuoteImport>,
    #[serde(rename = "overwriteExisting")]
    overwrite_existing: bool,
}

async fn import_quotes_csv(
    State(state): State<Arc<AppState>>,
    Json(body): Json<ImportQuotesBody>,
) -> ApiResult<Json<Vec<QuoteImport>>> {
    let result = state
        .quote_service
        .import_quotes(body.quotes, body.overwrite_existing)
        .await?;

    // Quote import - no market sync needed
    enqueue_portfolio_job(
        state,
        PortfolioJobConfig {
            account_ids: None,
            market_sync_mode: MarketSyncMode::None,
            force_full_recalculation: false,
        },
    );

    Ok(Json(result))
}

#[derive(serde::Deserialize)]
struct SyncBody {
    #[serde(rename = "assetIds")]
    asset_ids: Option<Vec<String>>,
    #[serde(rename = "refetchAll")]
    refetch_all: bool,
    #[serde(rename = "refetchRecentDays")]
    refetch_recent_days: Option<i64>,
}

async fn sync_market_data(
    State(state): State<Arc<AppState>>,
    Json(body): Json<SyncBody>,
) -> ApiResult<StatusCode> {
    // Determine the appropriate market sync mode based on refetch flags
    let market_sync_mode = if let Some(days) = body.refetch_recent_days {
        MarketSyncMode::RefetchRecent {
            asset_ids: body.asset_ids,
            days,
        }
    } else if body.refetch_all {
        MarketSyncMode::BackfillHistory {
            asset_ids: body.asset_ids,
            days: 365 * 5, // 5 years of history as fallback
        }
    } else {
        MarketSyncMode::Incremental {
            asset_ids: body.asset_ids,
        }
    };

    enqueue_portfolio_job(
        state,
        PortfolioJobConfig {
            account_ids: None,
            market_sync_mode,
            force_full_recalculation: false,
        },
    );
    Ok(StatusCode::NO_CONTENT)
}

#[derive(serde::Deserialize)]
#[serde(rename_all = "camelCase")]
struct LatestQuotesBody {
    asset_ids: Vec<String>,
}

async fn get_latest_quotes(
    State(state): State<Arc<AppState>>,
    Json(body): Json<LatestQuotesBody>,
) -> ApiResult<Json<std::collections::HashMap<String, Quote>>> {
    let quotes = state.quote_service.get_latest_quotes(&body.asset_ids)?;
    Ok(Json(quotes))
}

pub fn router() -> Router<Arc<AppState>> {
    Router::new()
        .route("/providers", get(get_market_data_providers))
        .route(
            "/providers/settings",
            get(get_market_data_provider_settings).put(update_market_data_provider_settings),
        )
        .route("/market-data/search", get(search_symbol))
        .route("/market-data/quotes/history", get(get_quote_history))
        .route("/market-data/quotes/latest", post(get_latest_quotes))
        .route("/market-data/quotes/{symbol}", put(update_quote))
        .route("/market-data/quotes/id/{id}", delete(delete_quote))
        .route("/market-data/quotes/import", post(import_quotes_csv))
        .route("/market-data/sync/history", post(sync_history_quotes))
        .route("/market-data/sync", post(sync_market_data))
}
